# Лабораторная работа №1
## Выполнил студент группы БФИ2001 Спица Илья Михайлович

### Оглавление
1. [Задание 1](#Задание-№1)
2. [Задание 2](#Задание-№2)
3. [Задание 3](#Задание-№3)
4. [Вывод](#Вывод)

### Задание №1
<i> Вызвать функцию print() и передать туда строку Hello, World! </i>

In [ ]:
print("Hello, World!")

### Задание №2
Написать генератор случайных матриц(многомерных), который принимает
опциональные параметры <b>m</b>, <b>n</b>, <b>min_limit</b>, <b>max_limit</b>, где <b>m</b> и <b>n</b> указывают размер
матрицы, а <b>min_lim</b> и <b>max_lim</b> - минимальное и максимальное значение для
генерируемого числа.

In [19]:
import random
import time

In [20]:
user_m = int(input("Введите количество строк "))
user_n = int(input("Введите количество столбцов "))
user_min_limit = int(input("Введите минимально возможное значение для генерации "))
user_max_limit = int(input("Введите максимально возможное значение для генерации "))

Matrix = [[random.randint(user_min_limit,user_max_limit) for j in range(user_n)] for i in range(user_m)]

Введите количество строк15000
Введите количество столбцов1
Введите минимально возможное значение для генерации1
Введите максимально возможное значение для генерации10


### Задание №3
Реализовать методы сортировки строк числовой матрицы в соответствии с
заданием. Оценить время работы каждого алгоритма сортировки и сравнить его со
временем стандартной функции сортировки. Испытания проводить на сгенерированных
матрицах.

In [21]:
import copy

In [22]:
# Сортировка выбором.
# Берётся срез массива, в котором минимальный элемент переносят в самый левый угол,
# после чего срез уменьшается и цикл повторяется.
start_time = time.time()
#TODO
def selection_sort(a):
    for i in range(0,len(a)):
        min = i;
        for j in range(i+1,len(a)):
            if a[j] < a[min]:
                min = j
        a[i],a[min] = a[min],a[i]
    return a

a1 = Matrix.copy()
start_time = time.time()
selection_sort(a1)
print("--- {0} ms ---".format(round((time.time() - start_time)*1000)))

--- 14236 ms ---


In [23]:
# Сортировка вставкой.
start_time = time.time()
#TODO
def insertion_sort(a):
    for i in range(1,len(a)):
        x = a[i]
        j = i
        while j>0 and a[j-1]>x:
            a[j]=a[j-1]
            j-=1
        a[j]=x
    return a

a2 = Matrix.copy()
start_time = time.time()
insertion_sort(a2)
print("--- {0} ms ---".format(round((time.time() - start_time)*1000)))

--- 16657 ms ---


In [24]:
# Сортировка обменом. || Сортировка пузырьком.
start_time = time.time()
#TODO
def psrk_sort(a):
    for i in range(len(a)-1):
        for j in range(len(a)-i-1):
            if a[j]>a[j+1]:
                a[j],a[j+1]=a[j+1],a[j]
    return a

a3 = Matrix.copy()
start_time = time.time()
psrk_sort(a3)
print("--- {0} ms ---".format(round((time.time() - start_time)*1000)))

--- 28103 ms ---


In [28]:
# Сортировка Шелла.
start_time = time.time()
#TODO
def sheell_sort(a):
    n = len(a)
    step = n//2
    while step>0:
        for i in range(step,n):
            j = i
            while j>=step and a[j]<a[j-step]:
                a[j],a[j-step] = a[j-step],a[j]
                j = j-step
        step = step//2
    return a

a4 = Matrix.copy()
start_time = time.time()
sheell_sort(a4)
print("--- {0} ms ---".format(round((time.time() - start_time)*1000)))

--- 67 ms ---


In [29]:
# Быстрая сортировка.
start_time = time.time()
#TODO
def quic_sort(a):
    if len(a) <= 1:
        return a
    else:
        p = random.choice(a)
        s_nums = []
        b_nums = []
        e_nums = []
        for n in a:
            if n < p:
                s_nums.append(n)
            elif n > p:
                b_nums.append(n)
            else:
                e_nums.append(n)
        return quic_sort(s_nums) + e_nums + quic_sort(b_nums)
    
a5 = Matrix.copy()
start_time = time.time()
quic_sort(a5)
print("--- {0} ms ---".format(round((time.time() - start_time)*1000)))
    

--- 13 ms ---


In [33]:
# Турнирная сортировка.
start_time = time.time()
#TODO
class PriorityHeap:
    def __init__(self) -> None:
        self.cache:list[init] = []
        self.size:int = 0
        self.max_size = 32
    def pop_max(self) -> any: #Удаление элементов - удаляем самый первый элемент 
        res = self.cache[0] #Самый маленьий первый элемент
        self.size -= 1
        self. cache[0] = self.cache[self. size] #В первый элемент записываем последний
        self. cache.pop(self.size)#Удаляем последний элемент
        self.iter_heapify(0)
        return res
    def max_heap_insert(self, key: int):
        self.cache. append(-10000000000000) #добавляем такое отриц. число, т.к в питоне нет максимального целого отриц. числа
        self.insert_key(key, self. size) #Тут в конец добавляем очень маленькое число и потом пытаемся на его место вставить наше новое значение
        self.size += 1
    def insert_key(self, value: int, i: int):
        self.cache [i] = value
        parent = self.parent(i)  #у нас есть i элемент, его родитель будет i/2-1
        while i > 0 and self.cache[parent] > self.cache[i]: #невозрастающее дерево, поэтому элементы уровня выше 
            # должны быть меньше, потому что на верхушке самое минимальное
            self.cache[parent], self. cache[i] = self.cache[i], self.cache[parent]
            i = parent
            parent = self.parent(i)
    def parent(self, i: int) -> int:
        res = (i >> 1) - 1
        if res < 0:
            return 0
        else:
            return res
    def iter_heapify(self, i: int) -> None: #Восстанавливает наше неубывающее дерево для восстановления порядка после того, как последний элемент закинули на первое место
        while True:
            l:int = (i << 1)+1 #Сдвиги - i/2
            r:int = (i << 1)+2
            largest:int = i
            if l < self.size and self.cache[l] < self.cache[largest]:
                largest = l
            if r < self.size and self. cache[r] < self.cache[largest]:
                largest = r
            if largest == i:
                break
            else:
                self.cache[i], self.cache[largest] = self.cache[largest], self.cache[i] 
                i = largest
def turnir_sort(a:list[int])->list[int]:
    a_len: int = len(a)
    heap = PriorityHeap()
    winners: list [int] = []
    losers: list[int] = []
    while heap.size < heap. max_size and heap.size < a_len: #пока пирамидка не заполнится и есть элементы в списке
        elem = a[heap.size]
        if winners and winners[-1] > elem:
            losers. append(elem)
        else:
            heap.max_heap_insert(a[heap.size])
    
    for i in range(heap.size, a_len):
        elem: int = a[i]
        if winners and winners[-1] > elem:
            losers. append(elem)
        else:
            max_elem : int = heap.pop_max()
            if not winners:
                winners.append(max_elem)
            elif winners[-1] <= max_elem:
                winners.append(max_elem)
            else:
                losers. append(max_elem)
                
            heap. max_heap_insert(elem)
    while heap.size > 0: #очищение пирамидки в победителей и лузеров
        max_elem: int = heap.pop_max()
        if not winners:
            winners.append(max_elem)
        elif winners[-1] <= max_elem:
            winners.append(max_elem)
        else:
            losers. append(max_elem)
    if not losers:
        return winners
    else:
        winners2: list[int] = turnir_sort(losers)
        i: int = 0
        j: int = 0 
        winners1_len:int = len(winners)
        winners2_len: int = len(winners2)  
        winners_res: list[int] = [] 
        while i < winners1_len and j < winners2_len:
            if winners[i] < winners2[j]:
                winners_res.append(winners[i])
                i += 1
            else:
                winners_res.append(winners2[j])
                j += 1
        while i < winners1_len:
            winners_res.append(winners[i])
            i += 1
        while j < winners2_len:
            winners_res.append(winners2[j])
            j += 1
        return winners_res
    
a6 = Matrix.copy()
start_time = time.time()
turnir_sort(a6)
print("--- {0} ms ---".format(round((time.time() - start_time)*1000)))

a7 = Matrix.copy()        
start_time = time.time()
a7.sort()                                             
print("--- {0} ms ---".format(round((time.time() - start_time))))

--- 77 ms ---
--- 0 ms ---


### Вывод


При вводе матрицы 15000х1 получились следующие результаты:
1. Сортировка выбором - 14236 ms
2. Сортировка вставкой - 16657 ms
3. Сортировка обменом - 28103 ms
4. Сортировка Шелла - 67 ms
5. Быстрая сортировка - 13 ms
6. Турнирная сортировка - 77 ms
7. Стандартная функция сортировки - 0 ms